### DISCUSSION TITLE
npm의 등장 배경과 사용함으로서 얻는 이점은 무엇이 있을까요?

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#90] [프론트 기술 질문 공간]
npm 등장 이전과 이후의 변화에 대해 알려주세요.

**Comment by michoball** [comment]
npm 등장 이전에는 프로젝트 개발에 필요한 패키지들을 일일히 다운을 받아야했습니다.  그리고 다른 개발자와 같은 개발환경을 공유하기 위해서는 내가 받은 의존성들을 동일하게 받아야했습니다. 이는 프로젝트 개발에 매우 번거롭고 에러가 생기기 쉬운 상황을 만들었을거라 생각됩니다. 
그런데 npm 이 생기고 난 후 npm과 package.json을 통해 다른 사람이 간편히 동일한 의존성들을 다운받아 사용할 수 있게 되었고
npm에 많은 사람들이 개발에 필요한 많은 크고 작은 라이브러리들을 배포하여 개발에 필요한 많은 라이브러리들을 손쉽게 사용할 수 있게 되었습니다. 
 이외에도 프로젝트에 대한 업데이트 버전을 체크할 수 있고, 사용자가 직접 사용하기 전에 실험적인 버전을 따로 사용해보거나 UX 를 해치지 않고 프로젝트를 업데이트를 할 수 있게 되었습니다.

**Comment by WooGie911** [comment]
먼저 npm은 패키지 레지스트리 및 의존성 관리 관리 툴입니다. 이는 글로벌 패키징이 아닌, 로컬 패키지를 사용하도록 하고, 패키지 버젼을 관리하고, 중복된 패키지를 관리해줍니다.

npm 등장 이전에는 하나의 글로벌 패키징을 할 경우 보통 한 가지 버젼만 사용할 수 있었는데 이는 여러 부작용을 촉발했습니다.
예를 들어, 여러 패키지가 의존하고 있는 하나의 모듈이 에러를 발생시키거나, 그 패키지의 API가 크게 바뀌거나, 악의적인 코드가 심겨져 있는 경우입니다. 

npm 은 여러 영리한 방법으로, 각 버젼의 패키지를 싹 다 설치하는 방향을 채택하였습니다. 
각 패키지마다 고유한 해쉬값을 부여햐여, 여러가지 버젼, 변형을 각각 설치하도록 하였으며, 이러한 경우 디스크 공간을 많이 잡아먹는다는 단점이 존재하지만, 다른 방법으로 최대한 디스크 공간 차지를 줄이려 합니다.


한 가지 방법으로 package.json에서 버젼을 살펴보면 틸다(~)와 캐럿(^) 표시, 부등호 표시들을 적극 활용하여 최대한 현재 키지를 재활용하려 합니다. 
이러한 방식을 시맨틱 버전 관리(Semantic Versioning)이라고 합니다.

**Comment by javascriptbiggosoo** [comment]
HTML 한 장으로 코딩을 수천수만줄 하다가 중간중간 필요한 라이브러리를 script로 받아오는 방식으로 코딩을 하면 무슨라이브러리를 어디서 받아왔는지, 의존성이 있는 라이브러리를 모두 같은 수준으로 업데이트했는지 파악하기 까다롭습니다.
스크립트를 가져오는 방식또한 매번 cdn 검색해서 복사해와야했죠.
npm을 쓰면 package.json에서 한 눈에 파악할 수 있게되고 스크립트를 다운받는것도 훨씬 간편해집니다.
